src/connectivity/mod.rs

```diff
@@ -628,6 +629,138 @@ pub fn graph_all_pairs_all_simple_paths(
     ))
 }
 
+/// Return a longest simple path in the graph
+///
+/// This function searches computes all pairs of all simple paths and returns
+/// a path of the longest length from that set. It is roughly equivalent to
+/// running something like::
+///
+///     from rustworkx import all_pairs_all_simple_paths
+///
+///     max((y.values for y in all_pairs_all_simple_paths(graph).values()), key=lambda x: len(x))
+///
+/// but this function will be more efficient than using ``max()`` as the search
+/// is evaluated in parallel before returning to Python. In the case of multiple
+/// paths of the same maximum length being present in the graph only one will be
+/// provided. There are no guarantees on which of the multiple longest paths
+/// will be returned (as it is determined by the parallel execution order). This
+/// is a tradeoff to improve runtime performance. If a stable return is required
+/// in such case consider using the ``max()`` equivalent above instead.
+///
+/// This function is multithreaded and will launch a thread pool with threads
+/// equal to the number of CPUs by default. You can tune the number of threads
+/// with the ``RAYON_NUM_THREADS`` environment variable. For example, setting
+/// ``RAYON_NUM_THREADS=4`` would limit the thread pool to 4 threads.
+///
+/// :param PyDiGraph graph: The graph to find the longest path in
+///
+/// :returns: A sequence of node indices that represent the longest simple graph
+///     found in the graph. If the graph is empty ``None`` will be returned instead.
+/// :rtype: NodeIndices
+#[pyfunction]
+#[pyo3(text_signature = "(graph, /)")]
+pub fn digraph_longest_simple_path(graph: &digraph::PyDiGraph) -> Option<NodeIndices> {
+    if graph.graph.node_count() == 0 {
+        return None;
+    } else if graph.graph.edge_count() == 0 {
+        return Some(NodeIndices {
+            nodes: vec![graph.graph.node_indices().next()?.index()],
+        });
+    }
+    let node_indices: Vec<NodeIndex> = graph.graph.node_indices().collect();
+    let node_index_set = node_indices.iter().copied().collect();
+    Some(NodeIndices {
+        nodes: node_indices
+            .par_iter()
+            .flat_map(|u| {
+                connectivity::all_simple_paths_multiple_targets(
+                    &graph.graph,
+                    *u,
+                    &node_index_set,
+                    0,
+                    None,
+                )
+                .values()
+                .flat_map(|paths| {
+                    paths
+                        .iter()
+                        .map(|path| path.iter().map(|v| v.index()).collect::<Vec<usize>>())
```

Comment:
I'm a bit confused by this line: wouldn't it be possible to already ask for the length of each path here? Or why do we need to iterate over that path and get the node indices (`v.index`)? (Maybe I also got lost and that's not what's actually happening 😛 )

Reply:
I've simplified the logic to minimize the intermediate object creation in https://github.com/Qiskit/rustworkx/pull/731/commits/f22f64b44077b292bd3eeaa0ffc780e323b7c185 and https://github.com/Qiskit/rustworkx/pull/731/commits/733e9ecd0e76a03e5a17412d8aebfb3908a7e66d I haven't benchmarked this it might be slightly faster (but I also expect the compiler was optimizing away a bunch in the previous version of the code)

Reply:
I played with it a little bit more and was able to drop all the copying so it should definitely be faster I think